runtime/vm/ExtendedMessageNPE.cpp

```diff
@@ -0,0 +1,201 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "cfreader.h"
+#include "stackwalk.h"
+#include "ut_j9vm.h"
+
+extern "C" {
+
+char*
+getCompleteNPEMessage(J9VMThread *vmThread, U_8 *bcCurrentPtr, J9ROMClass *romClass, const char *npeCauseMsg)
+{
+
+	char *npeMsg = NULL;
+	UDATA msgLen = 0;
+	const char *nlsMessage = NULL;
+	U_8 bcCurrent = *bcCurrentPtr;
+
+	PORT_ACCESS_FROM_VMC(vmThread);
+	Trc_VM_GetCompleteNPEMessage_Entry(vmThread, bcCurrentPtr, bcCurrent, romClass, npeCauseMsg);
+	if (((bcCurrent >= JBiaload) && (bcCurrent <= JBsaload))
+		|| ((bcCurrent >= JBiastore) && (bcCurrent <= JBsastore))
+	) {
+		const char *elementType = NULL;
+		switch (bcCurrent) {
+		case JBiaload: /* FALLTHROUGH */
+		case JBiastore:
+			elementType = "int";
+			break;
+		case JBlaload: /* FALLTHROUGH */
+		case JBlastore:
+			elementType = "long";
+			break;
+		case JBfaload: /* FALLTHROUGH */
+		case JBfastore:
+			elementType = "float";
+			break;
+		case JBdaload: /* FALLTHROUGH */
+		case JBdastore:
+			elementType = "double";
+			break;
+		case JBaaload: /* FALLTHROUGH */
+		case JBaastore:
+			elementType = "object";
+			break;
+		case JBbaload: /* FALLTHROUGH */
+		case JBbastore:
+			elementType = "byte/boolean";
+			break;
+		case JBcaload: /* FALLTHROUGH */
+		case JBcastore:
+			elementType = "char";
+			break;
+		case JBsaload: /* FALLTHROUGH */
+		case JBsastore:
+			elementType = "short";
+			break;
+		default:
+			Trc_VM_GetCompleteNPEMessage_Unreachable(vmThread, bcCurrent);
+		}
+		if ((bcCurrent >= JBiaload) && (bcCurrent <= JBsaload)) {
+#define VM_NPE_ALOAD "Cannot load from %s array"
+			nlsMessage = VM_NPE_ALOAD;
+#undef VM_NPE_ALOAD
```

Comment:
I'm not clear why this is using defines rather than the string directly?
```suggestion
			nlsMessage = "Cannot load from %s array";
```

Reply:
It was from an earlier commit in which there were multiple references to the string.
Fixed.